### PR TITLE
Fix & re-enable Travis CI

### PR DESCRIPTION
**Pull request by mrc0mmand**
This is a follow up PR for discussion from https://github.com/systemd/systemd/issues/10489#issuecomment-434838769. The main goal is to re-enable the Travis CI, right now just for Fedora + Coverity (with possibility of another distros in the future).

This is a WIP as I'd like to get some feedback before delving deeper. Right now it compiles systemd from given PR on Fedora Rawhide and runs `ninja test`.

Next steps:
- [x] Coverity support (example build log: https://travis-ci.org/mrc0mmand/systemd/builds/451081170)
- [ ] Integration testsuite (probably reuse some scripts from https://github.com/systemd/systemd-centos-ci)
- [ ] Support other distros (?)

The former `.travis.yml` configuration built a custom Docker image and uploaded it to the DockerHub - *imo* to deal with the volatile storage between stages. I'm not sure if it served any other purpose, but I just simply use the official Fedora Rawhide Docker image.

In the current state there is one test case from `ninja test` which keeps failing, but that's most likely caused by Docker misconfiguration:

```
The output from the failed tests:
298/499 test-capability                         FAIL     0.02 s (killed by signal 6 SIGABRT)
--- command ---
SYSTEMD_KBD_MODEL_MAP='/build/src/locale/kbd-model-map' PATH='/build/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' SYSTEMD_LANGUAGE_FALLBACK_MAP='/build/src/locale/language-fallback-map' /build/build/test-capability
--- stderr ---
have ambient caps: yes
Capabilities:= cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read+eip
Capabilities:= cap_dac_override,cap_net_raw+ep
Capabilities:= cap_dac_override+ep
Failed to drop auxiliary groups list: Operation not permitted
Failed to change group ID: Operation not permitted
Assertion 'fv == CAP_CLEAR' failed at ../src/test/test-capability.c:184, function test_update_inherited_set(). Aborting.
-------
```
Full build log: https://travis-ci.org/mrc0mmand/systemd/builds/450916240

### CODE REVIEW

File: travis-ci/fedora-build.reqs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+ gcc
```
Comment (evverx) [inline travis-ci/fedora-build.reqs]: I think it'd be better to install the build dependencies via `dnf build-dep systemd` and leave this file for non-standard dependencies like `libasan`.

File: .travis.yml [hunk 2]
```diff
@@ -1,77 +1,93 @@
...
-                  export UPLOAD_URL="https://scan.coverity.com/builds"
-                  export TOOL_ARCHIVE="/tmp/cov-analysis-${PLATFORM}.tgz"
-
-                # Get Coverity tool
-              - $CI_TOOL_DIR/get-coverity.sh
-              - TOOL_DIR="$(find $TOOL_BASE -type d -name 'cov-analysis*')"
-
-                # Export env variables for Coverity scan
+              # Encrypted COVERITY_SCAN_TOKEN env variable (travis encrypt)
+              - secure: "UNQLspT89GYWuVKFqW5W5RyqqnYg5RvX20IrNraOddhpdV9nhKBtozrfmhGXDGZwfHGWHt6g7YROlD/NIMvDvThVJIEYvSQiXCoo2zRrwkl2siET5MjPfRG8numiLq0KX47KGmyBJISJZCgDUdNGqqGwgf7AhDN78I3XtgqjFT1z0mGl8n0wiFpKPi7i3nECvF4Mk7xCCHqwByaq0z5G9NkVlOvP1EyCxwv3B6I5Umfch7ibp7iH44YnVXILK+yEry5dMuctYwYkDouR80ChEPQQ5fhhpO4++HJmFuSpfMTeCHpucAd2xwSUijejYeN/GNQ177GxSSk/8hRBGcuSK8T/WJ+KiuJPhZObV8mw+a6+qdQssWY4F9jya5ZKbZ/yTbxjtQ0m4AgtL28P9bEze8pLh16zFMX+hIEuoFSNmJqmtNttfbD5TKyYVZml59s9wvhlvMnlNpRSQva88OAOjXtiA41g+XtTxxpfW9mgd7HYhzSBs1efNiK7PfkANgve7KIYMAmCAqasgb1IIAyX7stOlJH06QOFXNH55PmJLkkKyL3SMQzgryMDWegU+XbS8t43r0x14WLuE7sc9JtnOr/G8hthFaMRp8xLy9aCBwyEIkEsyWa50VMoZDa3Spdb4r1CKBwcGdCbyE4rCehwEIznbfrsSovhwiUds7bbhBU="
```
Comment (evverx) [inline .travis.yml]: Just to be sure, did you pass `-r` to `travis encrypt`? If I recall correctly, by default, strings are encrypted with the kyes of the repository `origin` points to.

**Comment by evverx** [comment]
> I just simply use the official Fedora Rawhide Docker image

I think it's fine and, more importantly, judging by the logs, it seems to be working as opposed to the previous scheme (which was, in my opinion, also unnecessarily complicated).

In general, PR looks good to me and I don't think it's necessary to wait for all the points on the list to be implemented. We could think, for example, about https://docs.travis-ci.com/user/build-matrix/ when more than one "platform" is ready.

Anyway, the unit test should probably be fixed first to make Travis green.

File: .travis.yml [hunk 3]
```diff
@@ -1,77 +1,93 @@
...
+          install:
+              # Pull a Docker image and start a new container
+              - docker pull fedora:$FEDORA_RELEASE
+              - docker run -v $TRAVIS_BUILD_DIR:/build:rw
+                           -v /var/lib/dbus/machine-id:/etc/machine-id:ro
+                           -w /build --privileged=true --name $CONT_NAME
+                           -dit --net=host fedora:$FEDORA_RELEASE /sbin/init
+              # Upgrade system and install necessary build/test requirements
+              - $DOCKER_EXEC dnf -y --refresh upgrade
+              - $DOCKER_EXEC dnf -y install $(cat $BUILD_REQS_FILE)
```
Comment (evverx) [inline .travis.yml]: I decided to reproduce the test failure and gave up after `docker pull` :-), which got me thinking. Could you put these bits in a script and run the script instead? This way it'd be a little bit easier to create almost the same (hopefully :-)) environment locally (which I'm doing now to reproduce the test failure). In other words, it shouldn't be too hard for contributors to reproduce issues.

File: .travis.yml [hunk 4]
```diff
@@ -1,77 +1,93 @@
...
+          script:
+              - set -e
+              # Build systemd
+              - $DOCKER_EXEC meson build
+              - $DOCKER_EXEC ninja -C build
+              # Run 'make check'
+              - $DOCKER_EXEC ninja -C build test
+              - set +e
+
+        - stage: Coverity
```
Comment (evverx) [inline .travis.yml:39]: I think it would make sense to use something more stable than rawhide for Coverity.

File: .travis.yml (hunk 3 above)
Comment (evverx) [reply .travis.yml]: I gave up :-) Could you try passing `--security-opt seccomp:unconfined` to docker to see how it goes? If it doesn't help could you run the test under `strace` and paste the output here? Something like `strace -s500 -yy -v -f <path-to-test>` should suffice.

File: travis-ci/fedora-build.reqs (hunk 1 above)
Comment (mrc0mmand) [reply travis-ci/fedora-build.reqs]: Good idea, will fix.

File: .travis.yml (hunk 2 above)
Comment (mrc0mmand) [reply .travis.yml]: That's the original Coverity token, which *should* be the result of `travis encrypt -r systemd/systemd COVERITY_SCAN_TOKEN=xxx`

File: .travis.yml (hunk 3 above)
Comment (mrc0mmand) [reply .travis.yml]: That actually make sense, ~~we lose the phase splitting though (like `install`, `script`, ...). Simple `-e` should be a quick fix.~~

File: .travis.yml (hunk 4 above)
Comment (mrc0mmand) [reply .travis.yml:39]: Yeah, that was something I wanted to discuss. I guess using `fedora:latest`, which points to a latest, let's say *stable*, release, should be fine.

**Comment by mrc0mmand** [comment]
@evverx Unfortunately `--security-opt seccomp:unconfined` didn't help, here's the strace log: https://paste.fedoraproject.org/paste/-UajhkU~M9F-NyOZyS28vA

**Comment by mrc0mmand** [comment]
Latest batch of changes:

- moved the Docker magic to a separate script (`travis-ci/managers/fedora.sh`) which should make it easier to reproduce the environment for debugging

This took a little bit longer than I wanted as the `docker` package in Fedora is some weird Docker mutation and `dnf` has serious issues dealing with that. Long story short, you have to use the `docker-ce` package, e.g.:
```bash
dnf config-manager --add-repo https://download.docker.com/linux/fedora/docker-ce.repo
dnf install -y docker-ce
systemctl start docker
export CONT_NAME="my-fancy-container"
travis-ci/managers/fedora.sh SETUP RUN CLEANUP
```

You can run each phase separately and just play around with the container to you liking.

- removed the `travis-ci/fedora-build.reqs` file, as dealing with relative file includes in bash is *pita*; `dnf -y builddep systemd` and a short array of additional dependencies in the `travis-ci/managers/fedora.sh` is used instead
- Coverity check now uses `fedora:latest` image instead of `fedora:rawhide`, which should increase test results reliability.
- re-encrypted the `COVERITY_SCAN_TOKEN`, just in case

These changes should fix issues found by @evverx.

**Comment by evverx** [comment]
@mrc0mmand thank you! I'll try to take a look tomorrow.

> The dnf upgrade actually kills systemd

That certainly doesn't look right :-) I mean, it is fine to get around this temporarily by not installing problematic packages but `systemd` shouldn't crash every time it doesn't like the selinux policy or whatever it is that crashes it. Could you open an new issue?

**Comment by evverx** [comment]
> systemd shouldn't crash every time it doesn't like the selinux policy or whatever it is that crashes it

On the other hand, it's not clear what should happen in this case so probably freezing `systemd` is the safest option.

**Comment by mrc0mmand** [comment]
> On the other hand, it's not clear what should happen in this case so probably freezing `systemd` is the safest option.

File: travis-ci/managers/fedora.sh [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#
+# Example execution on Fedora:
+# dnf install docker
+# systemctl start docker
+# export CONT_NAME="my-fancy-container"
+# travis-ci/managers/fedora.sh SETUP RUN CLEANUP
+
+PHASES=(${@:-SETUP RUN CLEANUP})
+FEDORA_RELEASE="${FEDORA_RELEASE:-rawhide}"
+CONT_NAME="${CONT_NAME:-fedora-$FEDORA_RELEASE-$RANDOM}"
```
Comment (evverx) [inline travis-ci/managers/fedora.sh:14]: This doesn't seem to be working when `SETUP` and `RUN` are run separately. I got
```
FEDORA_RELEASE=latest ./travis-ci/managers/fedora.sh RUN
Run phase
Error response from daemon: No such container: fedora-latest-15116
```

File: travis-ci/managers/fedora.sh [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            $DOCKER_EXEC dnf makecache
+            # Install necessary build/test requirements
+            $DOCKER_EXEC dnf -y --exclude selinux-policy\* upgrade
+            $DOCKER_EXEC dnf -y install "${ADDITIONAL_DEPS[@]}"
+            $DOCKER_EXEC dnf -y builddep systemd
+            ;;
+        RUN)
+            info "Run phase"
+            # Build systemd
+            $DOCKER_EXEC meson build
```
Comment (evverx) [inline travis-ci/managers/fedora.sh:48]: This seems to fail when `meson` has been run in the current working directory:
```sh
Run phase
Directory already configured, exiting Meson. Just run your build command
(e.g. ninja) and Meson will regenerate as necessary. If ninja fails, run ninja
reconfigure to force Meson to regenerate.

If build failures persist, manually wipe your build directory to clear any
stored system data.

To change option values, run meson configure instead.
ninja: Entering directory `build'
[0/1] Regenerating build files.
Error during basic setup:

Neither directory contains a build file meson.build.
FAILED: build.ninja
/usr/bin/python3 /usr/bin/meson --internal regenerate /home/vagrant/systemd /home/vagrant/systemd/build --backend ninja
ninja: error: rebuilding 'build.ninja': subcommand failed
```

**Comment by evverx** [comment]
To sum up, `travis-ci/managers/fedora.sh` could be improved, but the script already makes it much easier to reproduce failures and as far as I can see, in its current form, it should work on Travis CI. So I still think that overwriting `test-capability` is the fastest way to make the CI work. Everything else could be sorted out later.

**Comment by evverx** [comment]
At last, I managed to run `RUN` :-) Another test that is failing in a container is `test-bpf`, which is exactly what https://github.com/systemd/systemd/issues/9666 is about :-) I guess it won't fail on Travis CI because the underlying kernel here is too old to support that feature.

**Comment by evverx** [comment]
By overwriting `test-capability` I mean something like
```diff
diff --git a/travis-ci/managers/fedora.sh b/travis-ci/managers/fedora.sh
index eed91d0d4..ab5ad154e 100755
--- a/travis-ci/managers/fedora.sh
+++ b/travis-ci/managers/fedora.sh
@@ -47,6 +47,7 @@ for phase in "${PHASES[@]}"; do
             # Build systemd
             $DOCKER_EXEC meson build
             $DOCKER_EXEC ninja -C build
+            $DOCKER_EXEC sh -c "printf '#!/bin/sh\necho The test is failing for unknown reason, skipping; exit 77' >/build/build/test-capability"
             # Run 'make check'
             $DOCKER_EXEC ninja -C build test
             ;;
```

**Comment by evverx** [comment]
@mrc0mmand I pushed the kludge on top of your branch to see how it goes.

**Comment by evverx** [comment]
Travis CI is green. I'm going to merge the PR to see whether Coverity still works.

**Comment by evverx** [comment]
Judging by https://travis-ci.org/systemd/systemd/jobs/451683385, the Coverity part seems to be broken:
```
travis_time:end:0e3a06d2:start=1541557596343939065,finish=1541557603786230863,duration=7442291798
[0K
[32;1mThe command "$DOCKER_EXEC sh -c "CFLAGS='-D_Float128=long\ double -D_Float64=double -D_Float64x=long\ double -D_Float32=float -D_Float32x=double' meson cov-build -Dman=false"" exited with 0.[0m
travis_time:start:1d294051
[0K$ $DOCKER_EXEC tools/coverity.sh build
[33;1mNote: COVERITY_SCAN_PROJECT_NAME and COVERITY_SCAN_TOKEN are available on Project Settings page on scan.coverity.com[0m
[33;1mCoverity Scan configured to run on branch master[0m
[33;1mCoverity Scan analysis authorized per quota.[0m
find: '/var/tmp/coverity-scan-analysis': No such file or directory
```
@mrc0mmand could you take a look?

**Comment by evverx** [comment]
I'm trying to fix it in https://github.com/systemd/systemd/pull/10664.

**Comment by evverx** [comment]
Coverity is back: https://travis-ci.org/systemd/systemd/builds/451717897; PRs are analyzed: https://travis-ci.org/systemd/systemd/pull_requests. Now we have a stable foundation to build on! Congratulations!

File: travis-ci/managers/fedora.sh (hunk 5 above)
Comment (mrc0mmand) [reply travis-ci/managers/fedora.sh:14]: That's somewhat expected, `export CONT_NAME="my-fancy-container"` from my example *fixes* it.

**Pull request merged**